WIP: ResponseSchema decorator

consense opened the pull request:
- Re #8
- new decorator added: `function ResponseBody(responseClass: any, statusCode: number = 200)`
- new test for above decorator added
- all tests passing locally 
- code linted
- BUT somehow code fails when including result of yarn run build into local project **even when not using the new decorator at all ;-(**

consense commented on src/decorators.ts:
```diff
@@ -52,3 +54,27 @@ function getOpenAPIMetadata(target: object, key: string): OpenAPIParam {
+ *
+ * @param responseType Class annotated with class-validator annotations
+ */
+export function ResponseBody(responseClass: any, statusCode: number = 200) {
```
unhappy about the "any" type for passed class. If you have some better idea how to ensure a class-validator-enabled or even a normal class is passed I'd be happy to learn something

epiphone commented:
Great, thanks for putting in the work!

> somehow code fails when including result of yarn run build into local project even when not using the new decorator at all ;-(

After a brief look I couldn't replicate this specific bug you encountered. I did give this another thought however and came up with other remarks: 

If at all possible I'd prefer to avoid importing `class-validator` and `class-validator-jsonschema` as non-development dependencies. The way the libraries work means that potential version clashes can have pretty nasty impacts, as demonstrated [here](https://github.com/typestack/class-validator/issues/166#issuecomment-362913525). Also, I image most people using this library already called the `validationMetadatasToSchemas` function in their app initialization code (and thus already parsed all the `class-validator` classes), so calling it again in each `ResponseObject` decorator might be redundant. An alternative would be to just to take the `ResponseObject` parameter class, use its name to build a schema reference string and assume that the user has already populated the `schemas` object in their spec. Simply passing a reference string instead of a class might work too.

Another thing I'm worried about is that since `ResponseBody` implementation takes a status code as an optional parameter, it might result in conflicts with `routing-controllers` `HttpCode` decorator. Also, the current implementation doesn't take into account any potential content-type decorators, but defaults to `json` instead. In general I think we should be careful to ensure the implementation plays well with existing decorators!

Also I think `ResponseSchema` might be more appropriate a name (better reflecting the scope of the decorator), especially if we end up getting rid of the status code parameter...  

Anyway, hope that makes any sense. Please let me know what you think!

consense commented:
Hey,

> If at all possible I'd prefer to avoid importing class-validator and class-validator-jsonschema as non-development dependencies. 
Yes makes sense, though thats maybe a case for `npm peerDependencies`? Ofc only if we really need the imports there anyway. I get your point regarding the assumption of all `class-validator` enabled classes having been parsed at this point so can gladly remove the dependencies for now.

> status code as an optional parameter, it might result in conflicts with routing-controllers HttpCode decorator.

Hmm - not sure. An action may have multiple response scenarios, so one would need to map the status code to the response-object. But yes - might just as well get rid of the default 200 and use the `HttpCode`  decorator value (and fallback to 200 otherwise same as routing-controllers with missing HttpCode. Only if someone feels the actual need to specify more distinctly might a parameter be necessary. YAGNI :-)
Do you have a suggestion on how to get the currently set `@HttpCode` (if any) from inside the decorator?
Same line of thought for the contentType - what would be the default case (we need to set the path inside the schema structure, so could react to `JsonController` instead of `Controller` being set (in addition to `@ContentType` decorator from RC.

> Also I think `ResponseSchema` might be more appropriate a name (better reflecting the scope of the decorator)

Gladly - no secret agendy for `ResponseBody` here

consense commented:
ah yes, the random failing I saw was probably due to me importing the locally-built RCO library wrongly somehow. Works find when declaring the new decorator inside application-land.

epiphone commented:
Setting status code as an optional secondary parameter and using @HttpCode decorator value (or 200 if that's not defined) as default works for me too. That's a good point about multiple response scenarios though. Supporting multiple `ResponseSchema` decorators as 

```typescript
@ResponseSchema(SomeObject)
@ResponseSchema(SomeErrorObject, 400)
handler() { ... }
```

or some sort of mapping like

```typescript
@ResponseSchema({200: SomeObject, 400: SomeErrorObject})
handler() { ... }
```
could be possible solutions. But I think these can be left for a later time.

> Do you have a suggestion on how to get the currently set @HttpCode (if any) from inside the decorator?

[Here's](https://github.com/epiphone/routing-controllers-openapi/blob/master/src/generateSpec.ts#L191) how we read `routing-controllers` status code and content type values. To access the `IRoute` object you can call OpenAPI [with a function](https://github.com/epiphone/routing-controllers-openapi/blob/master/src/decorators.ts#L11). I'll try to fix the problem of multiple `OpenAPI`s overwriting each other this week so calling it inside `ResponseSchema` won't be a problem.

consense commented:
- renamed decorator to `ResponseSchema`
- removed dependency on `class-validator` and `class-validator-jsonschema` as prod libs
- reading statusCode and contentType from RC decorators when possible
- optionally allowing to specify array response type
- 3 tests to cover the different way the statusCode/contentType could be set (RC decorator, options-hash, fallback-defaults)

disclaimer: I am personally compiling the openapi-v3 json back to swagger-v2, so while the output of the new decorator works quite well for me on https://www.screenaware.com/api , I didnt really end-to-end test the v3 version. In particular I am unsure whether the `[statusCode].content[contentType].schema` chain is correct this way in v3.

consense commented:
Hi, @epiphone Thanks for the feedback.
One point from my side before I do the adjustments to avoid another back-forth-itetation:
I am using the decorator in production and noticed the following problem:
If I am referencing a ResponseClass which is not class-validator decorated, the resulting openapi schema will not have the schema defined but the $ref set, so basically be inconsistent. In my case this led to a js error in swagger-ui when it tried to parse the spec. Of course it is easy to check for presence of the schema in the decorator and skip adding the ref in that case  but that would silently fail, leaving the developer wondering why the ResponseBody is not present in the spec.
Maybe console.log a warning?
What are your thoughts?

epiphone commented:
I wouldn't go for skipping the ref when schema is not present in decorator, since you might include the schema later after running `routingControllersToSpec`. I don't see a huge problem in the resulting spec turning out inconsistent when `ResponseClass` is a plain class as long as we clearly document that `ResponseClass` should be either class-validator decorated or manually included in the schema (maybe with the help of something like [`YousefED/typescript-json-schema`](https://github.com/YousefED/typescript-json-schema)). It's too bad that there's no way to check for that in `ResponseSchema` AFAIK, but on the other hand the resulting error in swagger-ui should be pretty understandable, right?

consense commented on src/decorators.ts:
```diff
@@ -55,3 +55,47 @@ function getOpenAPIMetadata(target: object, key: string): OpenAPIParam[] {
+ * Supplement action with response body type annotation.
+ *
+ */
+export function ResponseSchema(responseClass: Function, options?: {
```
I stand ashamed. Silly mistake

consense replied:
added including test

consense commented on __tests__/decorators.test.ts:
```diff
@@ -74,6 +76,32 @@ describe('decorators', () => {
+        return
+      }
     }
 
```
@epiphone hmm, not sure tbh. If I understand https://github.com/epiphone/class-validator-jsonschema/blob/master/src/decorators.ts correct,y JSONSchema will currently overwrite existing keys. So the only way to be "consistent" no matter the order in which the decorators are set, would be to be defensive in ResponseSchema and never overwrite. Is that what you ment?

epiphone replied:
As of #10 JSONSchemas are now `_.merge`d together instead of overwriting each other, so what I wrote on the above comment should work as intended (i.e. produce a `SomeResponseObject` response schema with `description: 'some description'`).

You're right about the order and consistency. In case of duplicate keys, for example

```typescript
@JSONSchema({responses: {
  '200': {
    'content': {
      'application/json': {
        schema: {
          '$ref': '#/components/schemas/Pet'
        }
      }
    ]
  }
}})
@ResponseSchema(SomeResponseObject)
handler() { ... }
```
the JSONSchema decorator overwrites ResponseSchema, and vice versa if their order was reversed. I think this is fine as long as we make it clear in the documentation.

consense commented on src/decorators.ts:
```diff
@@ -52,3 +54,27 @@ function getOpenAPIMetadata(target: object, key: string): OpenAPIParam {
+    responseSchema['' + statusCode]['application/json'].schema =
+      schemas[responseClass.name]
+  }
+  return OpenAPI({ responses: responseSchema })
```
Somehow couldnt reproduce this. I added 2 tests (OpenAPI->ResponseSchema and vice versa) and in both cases didnt notice one overwriting the other

epiphone replied:
Oh right, this was fixed in #10 so using `OpenAPI` isn't a problem anymore :+1:

consense commented on src/decorators.ts:
```diff
@@ -55,3 +55,47 @@ function getOpenAPIMetadata(target: object, key: string): OpenAPIParam[] {
+				}`;
+			}
+		}
+		return _.merge(source, { responses: responseSchema });
```
I'm quite ashamed. corrected now

consense commented on src/decorators.ts:
```diff
@@ -55,3 +55,47 @@ function getOpenAPIMetadata(target: object, key: string): OpenAPIParam[] {
+			},
+			...(options || {}),
+		};
+		const responseSchema = {
```
changed

consense commented on src/decorators.ts:
```diff
@@ -52,3 +54,27 @@ function getOpenAPIMetadata(target: object, key: string): OpenAPIParam {
+    responseClass.name &&
+    schemas[responseClass.name]
+  ) {
+    responseSchema['' + statusCode]['application/json'].schema =
```
true - changed

consense commented on __tests__/decorators.test.ts:
```diff
@@ -74,6 +76,32 @@ describe('decorators', () => {
+      responseSchemaDecorators(@Param('userId') _userId: number) {
+        return
+      }
+
```
added. Documentation had been added in previous commit, but clarified the order of priority a bit more

consense commented:
@epiphone regarding type of `responseClass` passed to :
`object` doesnt work as TS wont allow accessing `object.name` property to get the class name in that case. using `Function` works fine but collides with the tsline rules. Same for `any`. So for now I disabled tslint on the line using `// tslint:disable-next-line`.
Will be happy to learn something on how to solve this more elegantly.

consense commented:
done for now

consense commented:
Next round

epiphone commented:
Great, I'll try to get this published ASAP. Thanks for the contribution!
